### DISCUSSION TITLE
Using windowManager ShowMessageBox before Activation

### DISCUSSION THREAD

**Discussion by gmkado** [canton7/Stylet#207] [Q&A]
There are times during startup where something goes wrong and I want to get user input on how to proceed (think config file deserialization that doesn't parse correctly).  If I use `windowManager.ShowMessageBox` within the constructor the application closes prematurely.  

An easy repro is by inserting SayHello in the constructor of the `Stylet.Samples.Hello` example:

```
        public ShellViewModel(IWindowManager windowManager)
        {
            this.DisplayName = "Hello, Stylet";
            this.windowManager = windowManager;
            SayHello();
        }
```

After closing the message box, the application closes.  Not sure if this is by a design or if I'm doing something wrong with windowManager.

**Comment by canton7** [comment]
Interesting effect. ShellViewModel is constructed before its view is constructed, and the MessageBox happens to be the first window shown. This makes it WPF's main window, and so the application exits when it's closed.

I'm not honestly sure what I can do about this. ShellViewModel's ctor doesn't even return until the MessageBox has closed, so there's no way of grabbing an instance of ShellView and setting that as Application.MainWindow before the dialog is shown.

I don't think you can even go about setting Application.ShutdownMode = ShutdownMode.OnLastWindowClosed in the bootstrapper -- that has the same problem. 

Overall, I suspect the best approach is going to be not to write constructors which block for significant periods, and instead put that check in ShellViewModel.OnInitialActivate? That or, use ShutdownMode.Explicit, and explicitly shut down your application.

**Reply by gmkado** [reply]
Thanks for the advice, @canton7 

I ended up putting my `SettingsProvider.Load()` call in the bootstrapper and storing a flag to say whether it was successful or not, which I handle in `OnLaunch` after the main window is created.

```
        protected override void ConfigureIoC(IStyletIoCBuilder builder)
        {
             builder.Bind<ISettingsProvider>()
                .ToFactory(ioc =>
                {
                    // load the file before injection into other classes that need it
                    var sp = new SettingsProvider(Properties.Settings.Default.SettingFile);
                    try
                    {
                        sp.Load();
                    }
                    catch (Exception e)
                    {
                        SettingsLoadError = e.Message;
                    }
                    return sp;
                }) .InSingletonScope();
            }
           ...

        protected override void OnLaunch()
        {            
            if (!string.IsNullOrEmpty(SettingsLoadError)) {
                var result = Container.Get<IWindowManager>().ShowMessageBox(
                    $"Failed to load settings properly:\n{SettingsLoadError}.  Proceed with defaults?",
                    buttons: MessageBoxButton.YesNo);
                if (result == MessageBoxResult.No)
                {
                    Application.Shutdown();
                }
            } 
```
My reasoning for doing it this way as opposed to in `OnInitialActivate` is that I want the `ISettingsProvider.Settings` property to be set prior to injecting `ISettingsProvider` into other classes.  

Otherwise, all the classes that depend on it will be getting a null `ISettingsProvider.Settings` and will need to listen for the ISettings.PropertyChanged event to know when loading occurs, an implementation detail I didn't want to require/remember.

**Reply by canton7** [reply]
That seems sensible. You can also use `MessageBox.Show` from the bootstrapper if you need.
